CMakeLists.txt

```diff
@@ -0,0 +1,263 @@
+    /W4
+    /WX # All warnings as error
+    /we4013 # Error warning C4013: 'function' undefined; assuming extern returning int
+    /we4700 # Error warning C4700: uninitialized local variable
+    /we4715 # not all control paths return a value
+    /we4003 # not enough actual parameters for macro
+    /wd4456 # disable warning C4456: declaration of 'xxx' hides previous local declaration
+  >
+  $<$<NOT:$<C_COMPILER_ID:MSVC>>:
+    -Wall
+    -Wextra
+    -Wpedantic
+  >
+)
+
+target_compile_definitions(base64 PRIVATE
+  $<$<C_COMPILER_ID:MSVC>:
+    # remove unnecessary warnings about unchecked iterators
+    _SCL_SECURE_NO_WARNINGS
+  >
+)
+
+########################################################################
+# SIMD settings
+include(TargetSIMDInstructionSet)
+define_SIMD_compile_flags()
+
+if (_TARGET_ARCH STREQUAL "x86" OR _TARGET_ARCH STREQUAL "x64")
+    macro(configure_codec _TYPE)
+        if (BASE64_WITH_${_TYPE})
+            message(STATUS "Add codec: lib/arch/${_DIR}/codec.c")
+            string(TOLOWER "${_TYPE}" _DIR)
+            set_source_files_properties("lib/arch/${_DIR}/codec.c" PROPERTIES
+                COMPILE_FLAGS "${COMPILE_FLAGS_${_TYPE}}"
+            )
+
+            if (${ARGC} GREATER 1 AND MSVC)
+                set_source_files_properties("lib/arch/${_DIR}/codec.c" PROPERTIES
+                    COMPILE_DEFINITIONS ${ARGV1}
+                )
+            endif()
+        endif()
+    endmacro()
+
+    configure_codec(SSSE3 __SSSE3__)
+    configure_codec(SSE41 __SSSE4_1__)
+    configure_codec(SSE42 __SSSE4_2__)
+    configure_codec(AVX)
+    configure_codec(AVX2)
+
+elseif (_TARGET_ARCH STREQUAL "arm")
+    set(BASE64_NEON32_CFLAGS "${COMPILE_FLAGS_NEON32}" CACHE STRING "the NEON32 compile flags (for 'lib/arch/neon32/codec.c')")
+    mark_as_advanced(BASE64_NEON32_CFLAGS)
+
+    if (BASE64_WITH_NEON32)
+        set_source_files_properties("lib/arch/neon32/codec.c" PROPERTIES
+            COMPILE_FLAGS "${BASE64_NEON32_CFLAGS} "
+        )
+    endif()
+
+#elseif (_TARGET_ARCH STREQUAL "arm64" AND BASE64_WITH_NEON64)
+
+endif()
+
+configure_file("${CMAKE_CURRENT_LIST_DIR}/cmake/config.h.in" "${CMAKE_CURRENT_BINARY_DIR}/config.h" @ONLY)
+
+########################################################################
+# OpenMP Settings
+if (BASE64_WITH_OpenMP)
+    target_compile_options(base64
+        PRIVATE
+            ${OpenMP_C_FLAGS}
+    )
+endif()
+
+########################################################################
+if (BASE64_BUILD_TESTS)
+    enable_testing()
+    add_subdirectory(test)
+endif()
+
+########################################################################
+# cmake install
+install(DIRECTORY include/ TYPE INCLUDE)
+install(TARGETS base64 EXPORT base64-targets)
+
+include(CMakePackageConfigHelpers)
+configure_package_config_file(cmake/base64-config.cmake.in
+    "${CMAKE_CURRENT_BINARY_DIR}/base64-config.cmake"
+
+    INSTALL_DESTINATION "${CMAKE_INSTALL_DATADIR}/base64"
```

Comment:
cmake scripts are commonly installed to `${CMAKE_INSTALL_LIBDIR}/cmake/${PROJECT_NAME}`

Reply:
Have people finally settled on that convention? The [`install(EXPORT)`] and [`configure_package_config_file()`] docs don't explicitly recommend a destination, but provide examples with the following conflicting patterns:
- `$LIBDIR/${PROJECT_NAME}`
- `$LIBDIR/${PROJECT_NAME}/cmake`

Then of course there is the plethora other of options supported by [`find_package()`]. I usually opt for `$DATADIR/${PROJECT_NAME}` because that's the default for [`vcpkg_cmake_config_fixup()`] which in turn makes my [`portfile.cmake`] a little bit leaner.

However, I have no strong feelings one way or the other.

[`install(EXPORT)`]: https://cmake.org/cmake/help/latest/command/install.html?highlight=DESTINATION#export
[`configure_package_config_file()`]: https://cmake.org/cmake/help/latest/module/CMakePackageConfigHelpers.html?highlight=INSTALL_DESTINATION#example-generating-package-files
[`find_package()`]: https://cmake.org/cmake/help/latest/command/find_package.html#search-procedure
[`vcpkg_cmake_config_fixup()`]: https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/ports/vcpkg-cmake-config/vcpkg_cmake_config_fixup.md
[`portfile.cmake`]: https://github.com/deeplex-net/vcpkg-registry/blob/master/ports/base64/portfile.cmake

Reply:
It's just a matter of preference, I guess. Both are supported by [`find_package`](https://cmake.org/cmake/help/latest/command/find_package.html#search-procedure)
On my Fedora Linux systems, all cmake scripts, except one, are installed in `/usr/lib64/cmake`.


You can always make it configurable, and use the variable everywhere.
Then everybody is happy.
e.g.
```cmake
set(BASE64_INSTALL_CMAKEDIR "${CMAKE_INSTALL_DATADIR}/${PROJECT_NAME}" CACHE PATH "Where to install cmake scripts of base64")
```

Reply:
> It's just a matter of preference, I guess. [...]
> On my Fedora Linux systems, all cmake scripts, except one, are installed in `/usr/lib64/cmake`.

That's good enough for me and since there are no inherent technical drawbacks I'll go with your first suggestion; no need for unicorn solutions here.

> You can always make it configurable, and use the variable everywhere.

I'd like to avoid that if possible as the saying goes "All problems in computer science can be solved by another level of indirection -- except for the problem of too many layers of indirection." 😉